### PR TITLE
Fix restart of AppImage after update has finished

### PR DESCRIPTION
**Pull request by fmoc**
CC #11132.

The second commit cleans up the updater a bit.

### CODE REVIEW

File: src/gui/updater/ocupdater.cpp [hunk 1]
```diff
@@ -54,15 +54,7 @@ UpdaterScheduler::UpdaterScheduler(Application *app, QObject *parent)
...
-                connect(_updateDownloadedDialog, &UpdateDownloadedDialog::accepted, this, []() {
-                    // restart:
-                    qCInfo(lcUpdater) << "Restarting application NOW, PID" << qApp->applicationPid() << "is ending.";
-                    QTimer::singleShot(0, qApp->quit);
-                    QStringList args = qApp->arguments();
-                    QString prg = args.takeFirst();
-
-                    QProcess::startDetached(prg, args);
-                });
+                connect(_updateDownloadedDialog, &UpdateDownloadedDialog::accepted, this, []() { Utility::selfRestart(); });
```
Comment (TheOneRing) [inline src/gui/updater/ocupdater.cpp:57]: No need for a lambda

File: src/common/utility.cpp [hunk 2]
```diff
@@ -563,6 +563,53 @@ QString Utility::currentCpuArch()
...
+    }
+
+    for (const auto &arg : QCoreApplication::arguments().sliced(1)) {
+        args.push_back(strdup(arg.toStdString().c_str()));
+    }
+
+    // must be null-terminated, as we don't pass a length
+    args.push_back(nullptr);
+
+    execvp(pathToLaunch.toStdString().c_str(), args.data());
```
Comment (TheOneRing) [inline src/common/utility.cpp:602]: Well that might work with a simple application, but with this code the old instance was not properly shut down?

File: src/gui/updater/ocupdater.cpp [hunk 3]
```diff
@@ -563,6 +537,39 @@ void NSISUpdater::slotSetPreviouslySkippedVersion()
...
+{
+    Q_ASSERT(downloadState() == DownloadState::DownloadComplete);
+
+    auto settings = ConfigFile::makeQSettings();
+    QString updateFile = settings.value(updateAvailableC).toString();
+    settings.setValue(autoUpdateAttemptedC, true);
+    settings.sync();
+    qCInfo(lcUpdater) << "Running updater" << updateFile;
+
+    if (updateFile.endsWith(QLatin1String(".exe"))) {
```
Comment (TheOneRing) [inline src/gui/updater/ocupdater.cpp:550]: we can drop the support for nsis

File: src/common/utility.cpp (hunk 2 above)
Comment (fmoc) [reply src/common/utility.cpp:602]: Of course it is. `exec*` terminates the old application properly, i.e., all of Qt's shutdown code is run.

**Review by erikjv — APPROVED** [review]

**Comment by saw-jan** [comment]
Tested: `ownCloud-6.0.0.13582-daily20240327` -> `ownCloud-6.0.0.13589-daily20240328`

AppImage gets updated and restarted :heavy_check_mark: 

Screenshot from 2024-03-28 16-43-56 | Screenshot from 2024-03-28 16-47-25 | Screenshot from 2024-03-28 16-47-39
--|--|--

**Pull request merged**